samd: Support WiFi using external esp32 based modules.

robert-hh opened the pull request:
Such like the u.blox W102, Adafruit Airlift products or just any ESP32 based hardware. The ESP32-WROOM with at least 2MB flash module is sufficient. The actual firmware support full WiFi with SSL on boards with a SAMD51x19 MCU external flash and basic WiFi without SSL for SAMD51x19 boards with the file system in internal flash. SAMD21x18 boards with external flash could as well provide basic WiFi.
The build was tested using the test script and performs identical to the Arduino Nano 2040 connect, which is not a surprise. The PR includes documentation.
For providing the firmware binaries requires required for the ESP32 modules I have made PR to micropython-lib in the espflash path. It seemed to be the best place for it.
This extension is based on the great work of @iabdalkader with the network_ninaw10 package and his support during the implementation.

dpgeorge commented on docs/library/network.WLAN.rst:
```diff
@@ -15,13 +15,28 @@ This class provides a driver for WiFi network processors.  Example usage::
+    # The value for the GP0 pin can be set to `None`.
+    nic = network.WLAN(network.STA_IF, wiring=(SPI_obj, CS_pin, Busy_pin, RST_pin, GP0_pin))
+    # or if the GP0 pin is not wired, since it is not needed for WLAN
+    nic = network.WLAN(network.STA_IF, wiring=(SPI_obj, CS_pin, Busy_pin, RST_pin, None))
```
I discussed this with @jimmo.  To solve this problem properly will take a fair bit of effort (in API design) because it needs to support the general case of (for example) multiple WLAN drivers of different types, eg a board with both an external Nina-based module and cyw43-based module.  How would one specify these different drivers?  Maybe `network.NinaW10` and `network.CYW43`?  And how would they work together since the former doesn't use lwIP while the latter does?

All we really need to support here is the Adafruit Airlift, which can be done simply by a compile-time configuration and available as `network.WLAN()`.  Ie it doesn't need any arguments to configure it, it just uses the default settings which are fixed for the Airlift board.

If someone wants to use a Nina-based module (esp32) on some other samd-based board then it's pretty simple to make your own board definition and build it.  Then `network.WLAN()` will just work without any arguments because it's all configured at the C level.

In the future we can support the more general cases of dynamically (at runtime) instantiating WLAN interfaces of various types, but for now, to keep things moving and without getting bogged down in (long) API discussions, I suggest to just keep it simple and only allow the settings (SPI bus etc) be configured at compile time.

robert-hh replied:
> All we really need to support here is the Adafruit Airlift, which can be done simply by a compile-time configuration and available as network.WLAN(). Ie it doesn't need any arguments to configure it, it just uses the default settings which are fixed for the Airlift board.

It's not only Adafruit airlift, it's the general case. And even with Adafruit airlift there are various alternatives. There are:
- Boards with the airlift built-on, like the Metro M4 airlift. There, the wiring is fixed.
- Dedicated Add-on modules for the Itsy-Bitsy and Feather device, where the wiring is fixed as well.
- There is a UNO-Shaped Add-on board with fixed wiring, which works as well on some MIMXRT board, with a suggested wiring, which can be changed. I wired one up using a genuine ESP32 to be tested with MIMXRT boards. Works well at boards without Ethernet. If there is Ethernet, there is a conflict with the LWIP stack.
- And there are simple Adafruit and ESP32 breakout boards. I use one with a Sparkfun SAMD board. For these you could make a certain wiring mandatory. 
- SAMD boards for which no matching WiFi add.-on exists which still could use a simple ESP32 based breakout. For these boards a preferred wiring may not fit, just because the pins are not available or needed for a different purpose.

The way the code is implemented now, you can just write `network.WLAN()` if there is a wiring defined for the board and it is to be used. If one want to use a different type of connection, it must be specified as `network.WLAN(interface, spi, cs, busy, reset)` (GP0 is only required for Firmware update). So both options exist. And we are discussing here only about the NINAW10 network driver. So the question of having more than one WiFi is a different topic.

So I would suggest to (re-)add a preferred wiring to the SAMD board collection and keep the option of specifying an alternative wiring.

robert-hh replied:
@dpgeorge This update of the PR covers the default wiring schemes for the boards, where is is applicable, and I removed the changes to BLE() in the documentation. The Metro M4 Airlift PR had already the required defines, so it is not added here. The wiring for the Grand Central M4 Epress using the Airlift shield can be added as soon as the board files for this board is merged PR #11104.

dpgeorge replied:
> The way the code is implemented now, you can just write `network.WLAN()` if there is a wiring defined for the board and it is to be used. If one want to use a different type of connection, it must be specified as `network.WLAN(interface, spi, cs, busy, reset)`

I think this is close but needs a small change:
- If there is a default driver and wiring (eg NINA with specific pins like on the Airlift), then indeed `network.WLAN()` can be used
- Otherwise, one must be explicit with both the driver and the wiring.  I don't think it's a good idea to document that `network.WLAN()` can take a large set of keyword arguments, for all the different possible WLAN drivers.  Instead it would be more future proof to instantiate the driver from a class related to that driver, eg `network.NinaW10(...)`.

The documentation for `network.WLAN()` should not change.  Documentation for `network.NinaW10()` and its supported arguments (spi etc) would be added.

For boards like PYBD and Pico W the underlying WLAN driver would be `network.CYW43()` (although that class doesn't have to exist in the `network` module, `WLAN` can just be an alias to it, which is actually how it works at the moment).

If a samd board supports both a default WLAN and custom NINAW10 wiring, then it can have both `network.WLAN()` and `network.NinaW10()`.

------

Also, I don't think we should enable the NINAW10 driver by default on all the SAMD51 boards.  It adds a lot of code, and sets a precedent that's hard to back down from later on.  If you really want to provide pre-built firmware that has NINWA10 enabled then we can instead try adding board variants (like on stm32), so there is a base variant of the samd boards without networking, and a variant with networking included.

robert-hh replied:
> Documentation for network.NinaW10() and its supported arguments (spi etc) would be added.

Do you suggest adding network.NinaW10() as alias to network.WLAN(), so both can be used and the difference is only in the documentation?
___________________________
> Also, I don't think we should enable the NINAW10 driver by default on all the SAMD51 boards. It adds a lot of code, and sets a precedent that's hard to back down from later on. 

I agree that not all SAMD51 board can handle that, mechanically and by flash size. The size of the code is: Core WLAN drivers ~15k, MBEDTLS ~100k, network utility package ~10k. So the basic WLAN driver w/o encryption is small. And the code size may be less of a problem if support for external flash is present (PR #10233). But having variants is a good suggestion.

dpgeorge replied:
> Do you suggest adding network.NinaW10() as alias to network.WLAN(), so both can be used and the difference is only in the documentation?

In effect, yes that would be the change.

The idea is:
- If a board has built-in wlan then everyone just uses `network.WLAN()` to get at the interface and they don't need to think any further.
- If a board doesn't have built-in wlan (or lan, ble, etc) then one needs to think which driver they need and what connections to use.  The particular driver is specified by the class name.

robert-hh replied:
Fine. I'll set that up.

robert-hh commented:
@dpgeorge I hope I have implemented all of your suggestions right. The last commit tries to add variant support. Although it's SSL which weighs a lot.

robert-hh committed:
NINAW10: Add a network.NinaW10 class supporting external NINA modules.

It is possible to attach the network module shall be attached as
separate board. Therefore, the WLAN class is to be extended
allowing to define the connections. See separate PR.

- Add the NINA W10 drivers and networking

wlan = network.NinaW10(<ifmode>, *,
          spi=SPI, cs=CS_pin, busy=Busy_pin, reset=RST_pin)

The keywords are optional, unless only wiring is to be set. But then
all pins and spi must be provided.

All parameters must be objects of the respective types.
The signal names are those used by Adafruit airlift modules.

The constructor

wlan = network.WLAN(<ifmode>)

can still be used, if the default wiring for a board/WLAN module is
used.

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh committed:
samd/WLAN: Adapt the SAMD port files to support NINA base WLAN.

Including changes to the make and configuration files, and
changes to main.c and adding mbedtls config files.

The code area for samd51x20 devices is increased to 512k.

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh committed:
samd/WLAN: Change the board definition files for WLAN support.

Including:

ADAFRUIT_FEATHER_M4_EXPRESS
ADAFRUIT_ITSYBITSY_M4_EXPRESS
SPARKFUN_SAMD51_THING_PLUS

Pin definitions are added as well for:

ADAFRUIT_FEATHER_M0_EXPRESS
ADAFRUIT_ITSYBITSY_M0_EXPRESS

which can support basic WLAN once the use of the on-board flash for the
file system is available.

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh committed:
NINAW10: Add the network.NinaW10 class to the documentation.

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh commented:
@dpgeorge I reordered the commits into generic and samd-specific ones.

robert-hh commented:
Rebased to the actual master with external flash support. There is just one variant now, `wlan`. At boards with external flash that supports SSL, using mbtls. At boards without external flash no SSL. SAMD21 boards with external flash could be supported for WiFi without SSL, but it is not enabled yet. Only the pins are defined for two boards.

robert-hh commented:
Just a rebase and update after modules lost their "u".

robert-hh committed:
samd: Update the variant scheme.

- Remove the query_variant tag from Makefile.
- Remove the definition of BOARD_VARIANTS.

- Use uppercase for board variants.
- Append the variant name to the build directory name.

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh commented:
Updated because of the new board variant scheme.

I compared this implementation using the NINA firmware with a trial port using the esp_hosted firmware. The WiFi test coverage is the same. The timing behavior of the NINA firmware seems a little bit more robust. The major difference is at the code size. At BLE the esp_hosted firmware performs clearly better. Probably caused by the different esp-idf version. The NINA firmware uses 3.3.1, the esp_hosted firmware 5.0.2.

![nina-esp](https://github.com/micropython/micropython/assets/12476868/927c0c80-3538-4726-bd7a-bb2b3488094f)

iabdalkader commented:
> I compared this implementation using the NINA firmware with a trial port using the esp_hosted firmware.

First time that I notice this comparison, thank you!  It would be very interesting to see how they compare to CYW43.

robert-hh commented:
A short test for the CYW43:
WiFi test coverage good.
BLE test coverage good.
Bulk transfer speed **~1800 kByte/s** at >= 4 k block size !
Ping times from board to PC: 2 - 6 ms
Ping times from PC to board varying 1 - 2500ms

For esp-hosted:
Ping times from PC to board varying 63 - 250ms

For NINA W10:
Ping times from PC to board varying 24 - 120ms

iabdalkader commented:
> Bulk transfer speed **~1800 kByte/s** at >= 4 k block size !

That's very impressive.

robert-hh commented:
That's about 20Mbit/s. A better AP might even increase that. Mine is just a simple TP-Link unit. 
With Ethernet the Mimxrt boards get close to 100Mbit/s at their 100M interface.

iabdalkader commented:
Could look into improving the performance of the `esp_hosted`, specifically reading frames, I don't think we need to read the whole `ESP_FRAME_MAX_SIZE` maybe just enough to know the frame size. Also I think the firmware fragments packets over SPI that are less than the MTU.

robert-hh commented:
It seems there is always a full frame is sent. The screen shot below is just capturing the background activity. A block is 12800 clock cycles == 1600 bytes. The time between blocks is ~1 ms. The total speed is in any case limited by the SPI clock frequency in slave mode, which is 8Mhz for a standard ESP32.

![grafik](https://github.com/micropython/micropython/assets/12476868/976f11b4-2913-40d6-80e0-cfe3507b0e11)

robert-hh commented:
Another screen shot during bulk transfer. There seem to be many shorter packets of 868 bytes. I filled the data block with random numbers, so they appear as a solid block. But still there are not too many gaps between the blocks. Not what explains a factor of 10 in speed difference.

![grafik](https://github.com/micropython/micropython/assets/12476868/b59b8b0c-9027-4bf8-9d3f-ccf1b58aaa6c)

iabdalkader commented:
According to their benchmarks, the throughput should be much higher 

https://github.com/espressif/esp-hosted/blob/master/esp_hosted_fg/README.md#5-throughput-performance

Could it be the MCU ? How do I test this I can try it with the C33.

robert-hh commented:
I have seen the performance notes and was quite exited. The figures are close to the speed of the SPI interface.

The bulk transfer test script is pretty simple. See below. 10.0.0.152 is the IP address of the PC, 1234 the selected port. On the PC, run `nc -l 1234 >/dev/null`. Then run the script with:
```
import xmit
xmit.xmit(<no_of_blocks> [, <block_size_k>)
```
e.g. `xmit(1000, 4)`

```
import socket
from time import sleep_us, ticks_ms
import os

def xmit(count, blocks=1, delay=1000):
    blen = blocks * 1024
    b = bytearray(blen)
    for i in range(blen):
        b[i] = os.urandom(1)[0]
    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    s.connect(("10.0.0.152", 1234))
    mod = count // 100
    if mod < 10:
        mod = 10
    if mod > 10000:
        mod = 10000
    start = ticks_ms()
    sent = 0
    for i in range(count):
        try:
            s.sendall(b)
            sent += blen
            sleep_us(delay)
            if (i % mod) == 0:
                print(i)
        except Exception as e:
            print(i)
            print(e)
            break
    duration = ticks_ms() - start
    s.close()
    print(sent, "bytes sent", duration/count, "ms/frame", count * blen / (duration * 1000), "MB/s")
```

iabdalkader commented:
I will look into it, I was able to get close to 1Mbit by increasing TCP_MSS to 1600 from 800 (which might explain that ~800 bytes packets you've been seeing)
```
CLIENT MODE: TCP sending
Connecting to ('192.168.50.219', 5201)
Interval           Transfer     Bitrate
  0.00-1.00   sec   109 KBytes   895 Kbits/sec
  1.00-2.00   sec   105 KBytes   857 Kbits/sec
  2.00-3.00   sec   111 KBytes   912 Kbits/sec
  3.00-4.01   sec   110 KBytes   900 Kbits/sec
  4.01-5.01   sec   111 KBytes   910 Kbits/sec
  5.01-6.01   sec   109 KBytes   892 Kbits/sec
  6.01-7.01   sec   127 KBytes  1.04 Mbits/sec
  7.01-8.01   sec   102 KBytes   835 Kbits/sec
  8.01-9.01   sec   120 KBytes   983 Kbits/sec
  9.01-10.01  sec   125 KBytes  1.02 Mbits/sec
 10.01-10.05  sec  0.00 Bytes  0.00 bits/sec
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
  0.00-10.05  sec  1.10 MBytes   921 Kbits/sec  sender
  0.00-5.12   sec  1.10 MBytes  1.81 Mbits/sec  receiver
```

robert-hh commented:
That's what I see with the esp-hosted implementation. I get numbers between 130kByte/s to 200kByte/s, which is ~1 MBit/ to 1.6Mbit/s.

iabdalkader commented:
I used to see much lower rate, anyway, it seems to be directly related to the poll interval, if you lower it the rate goes higher.

robert-hh commented:
In the table of the tests the lower rates are about 60kByte/s for short data blocks.
I repeated the CYW43 test with the Pico W again, and now I see data rates of ~1350kBytes/s.  Maybe there is more traffic in the air now, like people streaming video.

iabdalkader commented:
I don't think the pin IRQ is working at all, I tried adding a printf there can't see it being called, might explain why lower the poll interval increases the throughput.

robert-hh commented:
I had a logic analyzer at the IRQ pin during tests, and never have seen any activity.

iabdalkader commented:
> I had a logic analyzer at the IRQ pin during tests, and never have seen any activity.

I thought it worked ?

Either way I think the wrong pin is used for IRQ, it should be data ready according to the docs:
```
#### Handshake pin
This is a output pin for ESP peripheral. ESP peripheral makes use of this pin to convey its readiness for execution of SPI transaction. The host is not supposed to initiate SPI transaction if ESP peripheral has not indicated its readiness. This pin stays high till the end of SPI transaction.

#### Data ready pin
This is a output pin for ESP peripheral. This pin is used to indicate host that the ESP peripheral wants to send a data packet to it. This pin stays high till the host reads this data packet.
```

Also this should be enough to indicate data ready:
```C
MP_WEAK bool esp_hosted_hal_data_ready(void) {
    return mp_hal_pin_read(MICROPY_HW_WIFI_DATAREADY);
}
```

I tried switching the IRQ pin to data ready, but I think the IRQ function is still not called.

robert-hh commented:
Sorry, I must have mixed that up with BLE. Below are sample activities with both the handshake and data ready pin. The time between DataReady going up and SPI activity is about 3 ms typically.
Tested with a SAMD51.

![grafik](https://github.com/micropython/micropython/assets/12476868/0db7971f-520d-47a0-8bf1-35ce318edd9d)

![grafik](https://github.com/micropython/micropython/assets/12476868/fc34ce2f-9e16-4252-969e-9cb645b5c5e3)

robert-hh commented:
Using a debugger with a breakpoint I can confirm that the IRQ is called.
At the Portenta C33 port, is `MICROPY_HW_WIFI_DATAREADY` of type Pin object? I see that it is defined as `pin_P803`. And in esp_hosted_hal.c, `MICROPY_HW_WIFI_IRQ` is defined as `MICROPY_HW_WIFI_HANDSHAKE`, which I do a well in the SAMD port. Setting it to MICROPY_HW_WIFI_DATAREADY makes things worse.

iabdalkader commented:
Yes that's right, it's `pin_P803` that pin is connected to IO9 which is the pin defined as Dataready, so it should be the pin used for IRQ not the handshake. Looking at the first capture above, it makes sense to use dataready not handshake. 

> Using a debugger with a breakpoint I can confirm that the IRQ is called.

I can't debug the C33 but I put a printf in there, not the best thing to do, but doesn't seem to be outputting anything. Btw reducing the poll interval to 1ms (just for testing) I get up to 2 or 3 Mbits.

iabdalkader commented:
Toggling an LED in the interrupt I see that the IRQ callback is called a few times when connecting, then it just stops, I think the IRQ is removed on soft-reboot, object is collected, or pins are deinitialized, or something like that.

robert-hh commented:
I usually just toggle a pin 8n a irq function printf is just too heavy.

iabdalkader commented:
> I usually just toggle a pin 8n a irq function printf is just too heavy.

It was removed/disabled after the first soft-reboot like I said. I have it working now steadily with some small changes, we just need to reinstall the IRQ from `esp_hosted_wifi_init()`. I don't know if it's the same in SAMD/MIMXRT but it's definitely getting removed in `ra`. With the IRQ fixed, I do see small improvements in performance, without increasing the poll interval, but nowhere near ESP's benchmark.

robert-hh commented:
SAMD and MIMXRT ports call `esp_hosted_wifi_deinit()` on soft reboot. That is needed since the SPI (and UART) objects are dynamic. So `esp_hosted_wifi_init()` has to be run again. Still I get better performance figures using the interrupt on `MICROPY_HW_WIFI_HANDSHAKE`. Dropping the poll timer to 32 ms increases the performance slightly.

iabdalkader commented:
> SAMD and MIMXRT ports call esp_hosted_wifi_deinit() on soft reboot.

That makes sense. I will need to fix the IRQ so it gets re-enabled if deinit is not called (like in `ra`) later. I also noticed that in RA an exact 30MHz clock is not possible, it should be 25MHz.

> Still I get better performance figures using the interrupt on MICROPY_HW_WIFI_HANDSHAKE.

This is likely because the handshake toggles more often to indicate that the host is ready for a new transfer (write) so it just schedules a poll more often, but it's data ready that should be the IRQ.

robert-hh commented:
Calling the IRQ on both events results in even better & stable performance. With SAMD51 I'm now at stable ~220kByte/s (~1.8Mbit/s) at 16k block size, whereas before it was between 100 and 160 kByte/s.

robert-hh commented:
Note: "If it does not work, take a larger hammer".

iabdalkader commented:
If you enable debugging by setting `ESP_HOSTED_DEBUG` 1 in hah.h (note you may need to remove ESP_HOSTED_DEBUG that's in internal.h, that's redefined and shouldn't be there) do you see this message :

```
esp_hosted_poll() invalid frame size 0 offset 0
```

It seems the first frame after data ready is always zeros.

robert-hh commented:
Yes. I've seen these before, and you suggested to ignore it.

iabdalkader commented:
> Yes. I've seen these before, and you suggested to ignore it.

I thought they were local SPI timing issue maybe, but seems they are something else.

robert-hh commented:
If you look at the burst below during data transfer, you can see that one block between A1 and A2 takes 32ms and has 93952 clock cycles, or at 8MHz 11.74ms net SPI time, or 36% net usage. If one looks at the SPI transfer itself, it has 68% net usage. I doubt that this can be squeezed much more. This is with TCP, and there is always protocol overhead both by TCP and WiFi.

![grafik](https://github.com/micropython/micropython/assets/12476868/5ccaaa94-d04f-4cb0-9597-20fa4d9febb6)

iabdalkader commented:
Maybe but note that I think that every other frame is empty, so half of those are wasted time.

robert-hh commented:
With IRQ at both signals, Ping from board to PC is more reliable and drops to ~17ms average.

robert-hh commented:
> Maybe but note that I think that every other frame is empty, so half of those are wasted time.

Do you see a way to get rid of the empty frames?

iabdalkader commented:
> Do you see a way to get rid of the empty frames?

I'm looking into it, I thought it was maybe an SPI issue in the port, but if you see those too then it's likely driver/firmware issue.

iabdalkader commented:
@robert-hh might explain the issue https://github.com/espressif/esp-hosted/issues/27#issuecomment-823331635

robert-hh commented:
Looks that we got the same figures. At MIMXRT1020 and IRQ at both dataready and handshake, I get a stable ~200kByte/s transfer rate and reliable ping times of 12-20 ms.

iabdalkader commented:
Whoa UDP seems much faster:
```
CLIENT MODE: UDP sending
Connecting to ('192.168.50.219', 5201)
Interval           Transfer     Bitrate         Total Datagrams
  0.00-1.00   sec   547 KBytes  4.47 Mbits/sec  384
  1.00-2.00   sec   548 KBytes  4.48 Mbits/sec  385
  2.00-3.01   sec   547 KBytes  4.47 Mbits/sec  384
  3.01-4.01   sec   548 KBytes  4.48 Mbits/sec  385
  4.01-5.01   sec   547 KBytes  4.47 Mbits/sec  384
  5.01-6.01   sec   548 KBytes  4.48 Mbits/sec  385
  6.01-7.01   sec   541 KBytes  4.43 Mbits/sec  380
  7.01-8.01   sec   548 KBytes  4.48 Mbits/sec  385
  8.01-9.02   sec   547 KBytes  4.47 Mbits/sec  384
  9.02-10.00  sec   535 KBytes  4.46 Mbits/sec  376
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
  0.00-10.00  sec  5.33 MBytes  4.47 Mbits/sec  3832  sender
  0.00-6.00   sec  5.32 MBytes  7.43 Mbits/sec  3832  receiver
```

robert-hh commented:
looks good.

robert-hh commented:
MIMXRT 1020

```
CLIENT MODE: TCP sending
Connecting to ('10.0.0.152', 5201)
Interval           Transfer     Bitrate
  0.00-1.00   sec   204 KBytes  1.67 Mbits/sec
  1.00-2.02   sec   215 KBytes  1.74 Mbits/sec
  2.02-3.03   sec   209 KBytes  1.70 Mbits/sec
  3.03-4.03   sec   202 KBytes  1.66 Mbits/sec
  4.03-5.04   sec   203 KBytes  1.64 Mbits/sec
  5.04-6.04   sec   198 KBytes  1.62 Mbits/sec
  6.04-7.05   sec   198 KBytes  1.62 Mbits/sec
  7.05-8.05   sec   198 KBytes  1.62 Mbits/sec
  8.05-9.06   sec   192 KBytes  1.55 Mbits/sec
  9.06-10.02  sec   188 KBytes  1.60 Mbits/sec
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
  0.00-10.03  sec  1.96 MBytes  1.64 Mbits/sec  sender
  0.00-23.04  sec  1.96 MBytes   714 Kbits/sec  receiver
```

```
CLIENT MODE: UDP sending
Connecting to ('10.0.0.152', 5201)
Interval           Transfer     Bitrate         Total Datagrams
  0.00-1.00   sec   461 KBytes  3.77 Mbits/sec  324
  1.00-2.00   sec   461 KBytes  3.77 Mbits/sec  324
  2.00-3.01   sec   461 KBytes  3.77 Mbits/sec  324
  3.01-4.01   sec   460 KBytes  3.76 Mbits/sec  323
  4.01-5.01   sec   460 KBytes  3.76 Mbits/sec  323
  5.01-6.01   sec   460 KBytes  3.76 Mbits/sec  323
  6.01-7.01   sec   461 KBytes  3.77 Mbits/sec  324
  7.01-8.02   sec   460 KBytes  3.76 Mbits/sec  323
  8.02-9.02   sec   458 KBytes  3.75 Mbits/sec  322
  9.02-10.00  sec   453 KBytes  3.77 Mbits/sec  318
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
  0.00-10.00  sec  4.49 MBytes  3.76 Mbits/sec  3228  sender
  0.00-16.00  sec  4.49 MBytes  2.35 Mbits/sec  3228  receiver
```

SAMD51

```
CLIENT MODE: TCP sending
Connecting to ('10.0.0.152', 5201)
Interval           Transfer     Bitrate
  0.00-1.03   sec   198 KBytes  1.58 Mbits/sec
  1.03-2.06   sec   191 KBytes  1.51 Mbits/sec
  2.06-3.08   sec   188 KBytes  1.51 Mbits/sec
  3.08-4.08   sec   215 KBytes  1.76 Mbits/sec
  4.08-5.08   sec   221 KBytes  1.81 Mbits/sec
  5.08-6.08   sec   220 KBytes  1.80 Mbits/sec
  6.08-7.09   sec   226 KBytes  1.83 Mbits/sec
  7.09-8.11   sec   219 KBytes  1.76 Mbits/sec
  8.11-9.13   sec   220 KBytes  1.77 Mbits/sec
  9.13-10.02  sec   194 KBytes  1.77 Mbits/sec
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
  0.00-10.02  sec  2.04 MBytes  1.71 Mbits/sec  sender
  0.00-10.79  sec  2.04 MBytes  1.59 Mbits/sec  receiver
```

```
CLIENT MODE: UDP sending
Connecting to ('10.0.0.152', 5201)
Interval           Transfer     Bitrate         Total Datagrams
  0.00-1.00   sec   356 KBytes  2.91 Mbits/sec  250
  1.00-2.00   sec   357 KBytes  2.92 Mbits/sec  251
  2.00-3.01   sec   356 KBytes  2.90 Mbits/sec  250
  3.01-4.01   sec   357 KBytes  2.92 Mbits/sec  251
  4.01-5.01   sec   352 KBytes  2.88 Mbits/sec  247
  5.01-6.01   sec   359 KBytes  2.93 Mbits/sec  252
  6.01-7.02   sec   353 KBytes  2.88 Mbits/sec  248
  7.02-8.02   sec   357 KBytes  2.92 Mbits/sec  251
  8.02-9.02   sec   355 KBytes  2.90 Mbits/sec  249
  9.02-10.00  sec   353 KBytes  2.93 Mbits/sec  248
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
  0.00-10.01  sec  3.47 MBytes  2.91 Mbits/sec  2497  sender
  0.00-10.38  sec  3.47 MBytes  2.81 Mbits/sec  2497  receiver
```

robert-hh commented:
Using the attached version of esp_hosted_hal.c, I get the following times using a MIMXRT1020:
```
>>> iperf3.client("10.0.0.152")
CLIENT MODE: TCP sending
Connecting to ('10.0.0.152', 5201)
Interval           Transfer     Bitrate
  0.00-1.00   sec   271 KBytes  2.22 Mbits/sec
  1.00-2.00   sec   263 KBytes  2.15 Mbits/sec
  2.00-3.01   sec   277 KBytes  2.26 Mbits/sec
  3.01-4.01   sec   263 KBytes  2.15 Mbits/sec
  4.01-5.01   sec   271 KBytes  2.22 Mbits/sec
  5.01-6.01   sec   274 KBytes  2.24 Mbits/sec
  6.01-7.01   sec   269 KBytes  2.19 Mbits/sec
  7.01-8.01   sec   267 KBytes  2.19 Mbits/sec
  8.01-9.02   sec   259 KBytes  2.11 Mbits/sec
  9.02-10.00  sec   266 KBytes  2.22 Mbits/sec
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
  0.00-10.00  sec  2.62 MBytes  2.19 Mbits/sec  sender
  0.00-21.93  sec  2.62 MBytes  1.00 Mbits/sec  receiver
>>> iperf3.client("10.0.0.152", udp=True)
CLIENT MODE: UDP sending
Connecting to ('10.0.0.152', 5201)
Interval           Transfer     Bitrate         Total Datagrams
  0.00-1.00   sec   682 KBytes  5.58 Mbits/sec  479
  1.00-2.00   sec   681 KBytes  5.56 Mbits/sec  478
  2.00-3.00   sec   679 KBytes  5.56 Mbits/sec  477
  3.00-4.00   sec   679 KBytes  5.56 Mbits/sec  477
  4.00-5.01   sec   678 KBytes  5.55 Mbits/sec  476
  5.01-6.01   sec   679 KBytes  5.55 Mbits/sec  477
  6.01-7.01   sec   681 KBytes  5.57 Mbits/sec  478
  7.01-8.01   sec   681 KBytes  5.57 Mbits/sec  478
  8.01-9.01   sec   681 KBytes  5.57 Mbits/sec  478
  9.01-10.00  sec   672 KBytes  5.56 Mbits/sec  472
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
  0.00-10.00  sec  6.63 MBytes  5.56 Mbits/sec  4770  sender
  0.00-16.23  sec  6.63 MBytes  3.43 Mbits/sec  4770  receiver
```
Especially the values for UDP are great.
[esp_hosted_hal.zip](https://github.com/micropython/micropython/files/12669167/esp_hosted_hal.zip)

iabdalkader commented:
@robert-hh I can achieve similar or slightly better results by adding this to the end of the hal's spi_transfer function

```
    if (esp_hosted_hal_data_ready()) {
        mod_network_poll_events();
    }
```
If dataready is still high, there are more packets so schedule another poll.

```
CLIENT MODE: TCP sending
Connecting to ('192.168.50.219', 5201)
Interval           Transfer     Bitrate
  0.00-1.02   sec   163 KBytes  1.31 Mbits/sec
  1.02-2.02   sec   158 KBytes  1.29 Mbits/sec
  2.02-3.02   sec   216 KBytes  1.77 Mbits/sec
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
  0.00-10.01  sec  1.66 MBytes  1.39 Mbits/sec  sender
  0.00-5.05   sec  1.66 MBytes  2.76 Mbits/sec  receiver
CLIENT MODE: UDP sending
Connecting to ('192.168.50.219', 5201)
Interval           Transfer     Bitrate         Total Datagrams
  0.00-1.00   sec   545 KBytes  4.47 Mbits/sec  383
  1.00-2.00   sec   541 KBytes  4.42 Mbits/sec  380
  2.00-3.00   sec   538 KBytes  4.40 Mbits/sec  378
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
  0.00-10.00  sec  5.05 MBytes  4.23 Mbits/sec  3631  sender
  0.00-5.59   sec  5.05 MBytes  7.58 Mbits/sec  3631  receiver
```

Note our sender/receiver results seem to be reversed.

robert-hh commented:
> If dataready is still high, there are more packets so schedule another poll.

Trying that I get the same figures for UDP, but 4 times lower data rates for TCP.

> Note our sender/receiver results seem to be reversed.

There is something strange happening with iperf3. It looks as if for a while one direction does not work. With SAMD51, both figures are the same.

iabdalkader commented:
> Trying that I get the same figures for UDP, but 4 times lower data rates for TCP.

With polling after SPI transfer (combined with some other changes)  I get:

- TCP 1.78/3.54 Mbits/sec  UDP 4.45/7.98 Mbits/sec

Without polling after SPI transfer:
- TCP 0.65/1.28 Mbits/sec  UDP 4.47/8.13 Mbits/sec

robert-hh commented:
Did you try the version with IRQ both at dataready and handshake?
Here with MIMXRT, polling after SPI transfer improves UDP, but not TCP.

iabdalkader commented:
> Did you try the version with IRQ both at dataready and handshake?

Yes and with same results more or less, enabling IRQ on both pins at this point just results in more unnecessary  polling.

robert-hh commented:
OK, For MIMXRT and SAMD it makes a big difference, so I keep it there. The amount of unneccessary polling seems minor. But there are sequence where several SPI tansfers happen with dataready low. I kept the periodic polling at 128ms, since for any immediate response polling is triggered by the IRQ. I do not know if you changed the polling time.

iabdalkader commented:
https://github.com/micropython/micropython/pull/12490

iabdalkader commented:
Those are all of the changes I have here https://github.com/micropython/micropython/pull/12490

> But there are sequence where several SPI tansfers happen with dataready low

If there's anything to read DATAREADY will be high already and spi_transfer will reschedule the poll as long as it's high.

robert-hh commented:
With all relevant changes from #12940 to esp_hosted_hal.c applied, the figures are the same as with the dual IRQ, under the varying environmental situation. The traffic pattern at the wires looks comparable.

I see that you still have support for a WiFi LED, if enabled.

iabdalkader commented:
> With all relevant changes from https://github.com/micropython/micropython/pull/12490 to esp_hosted_hal.c applied, the figures are the same as with the dual IRQ,

Yes it should be the same, but this doesn't require configuring the second pin. I think this is the max it can do, I'm running the SPI clock at 25MHz, unless there's something to optimize at the firmware side.

 > I see that you still have support for a WiFi LED, if enabled.

Yes it's a very good indicator, it blinks much faster when running iperf, also the IRQ pin is now configured in board config files.

robert-hh commented:
Updated the comparison of the boards.
![grafik](https://github.com/micropython/micropython/assets/12476868/ffe5c53c-8164-4aa1-8cb0-964a020d8c15)

robert-hh commented:
After some testing I made a small change to `esp_hosted_hal_spi_transfer()`, which improved ping reliability. The changes consists of increasing the second delay after setting CS to 1 before the repeated poll from 10µs to 100µs, at least for SAMD51. Other boards may works with smaller values. Before, there were about 2% ping events lost. After, I had not drop in several thousand attempts. The total performance dropped, if at all,  just a little bit.

```
    mp_hal_delay_us(100);  // Increased this one

    if (esp_hosted_hal_data_ready()) {
        void mod_network_poll_events(void);
        mod_network_poll_events();
    }
```

iabdalkader commented:
Yes I noticed it needs a small delay when doing repeated transfers, but I feel those minor fixes here and there will only get us so far, there's a bottleneck somewhere that's capping the TCP performance.

robert-hh commented:
This change is not about speed, it is about reliability. In my local LAN, ping should always succeed with a small variation of response times. That's what troubled me. The other figures are fine. 

10µs delay: Ping times ~10ms to sometimes >500ms, ~2% loss.
100µs delay: Ping times ~10ms to max ~30ms, 0 % loss.

iabdalkader commented:
No problem, I can add it to the PR.

robert-hh commented:
On a related topic: reducing the wait time in the polling loops of `nina_bsp_spi_slave_select()` for the NINAW10 solution from 1ms to e.g. 100µs nearly doubles the speed and reduces average ping times from 40ms to 20ms. Still a little bit slower than the esp_hosted solution, but not much.

iabdalkader commented:
> On a related topic: reducing the wait time in the polling loops of `nina_bsp_spi_slave_select()` for the NINAW10 solution from 1ms to e.g. 100µs 

Feel free to fix it in your PR 👍🏼

robert-hh commented:
Of course. I'm still not sure whether I should keep the new NINAW10 interface or just require the wiring to be fixed in the respective mpconfigboard.h files.

robert-hh commented:
> Feel free to fix it in your PR 👍🏼

I did, but changed the approach using the esp_hosted style of esp_hosted_hal.c, by declaring the functions in nina_wifi_bsp.c as `MP_WEAK` and allowing a port to replace them. So in the template, the delay stays at 1ms, which works better for the arduino_nano_2040, and for SAMD it's reduced to 100us. That file implements as well the SAMD-specific set-up for SPI. I dropped the generic network.NINAW10 interface.

iabdalkader commented:
Those are good changes, thank you!

iabdalkader commented:
BTW was able to get a higher throughput by increasing lwip's memory (SND_BUF etc..) but it takes a lot of memory:

```
CLIENT MODE: TCP sending
Connecting to ('192.168.50.219', 5201)
Interval           Transfer     Bitrate
  0.00-1.00   sec   354 KBytes  2.90 Mbits/sec
  1.00-2.04   sec   421 KBytes  3.33 Mbits/sec
  2.04-3.04   sec   369 KBytes  3.02 Mbits/sec
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
  0.00-10.00  sec  3.94 MBytes  3.30 Mbits/sec  sender
  0.00-5.14   sec  3.94 MBytes  6.43 Mbits/sec  receiver
CLIENT MODE: UDP sending
Connecting to ('192.168.50.219', 5201)
Interval           Transfer     Bitrate         Total Datagrams
  0.00-1.00   sec   528 KBytes  4.32 Mbits/sec  371
  1.00-2.00   sec   530 KBytes  4.33 Mbits/sec  372
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
  0.00-10.00  sec  4.85 MBytes  4.07 Mbits/sec  3487  sender
  0.00-5.45   sec  4.85 MBytes  7.46 Mbits/sec  3487  receiver
```

robert-hh commented:
I know that the LWIP settings are a dial to turn, but the increase is slow. You spend a lot of memory for a little bit higher speed. For the 8Mbit limit you should look into the net SPI transfer pattern. At the SAMD51 I see a short break between blocks of each 3 bytes, reducing already the net SPI transfer speed e.g. from 8Mbit to ~5.5MBit.

iabdalkader commented:
> You spend a lot of memory for a little bit higher speed

Actually I see significant throughput when increase SND_BUF + MTU etc.. even with CYW43, the defaults in MicroPython are too low and cause a lot of fragmentation (I guess the assumption is you wouldn't be sending large packets normally like iperf does).

BTW I also tested CYW43/RPI Pico_W (SPI mode) with the same LWIP config I use in my PR, and receiving is very good:
```
CLIENT MODE: TCP sending
Connecting to ('192.168.50.219', 5201)
Interval           Transfer     Bitrate
  0.00-10.00  sec  13.3 MBytes  11.1 Mbits/sec  sender
  0.00-10.25  sec  13.3 MBytes  10.9 Mbits/sec  receiver
CLIENT MODE: UDP sending
Connecting to ('192.168.50.219', 5201)
Interval           Transfer     Bitrate         Total Datagrams
  0.00-10.00  sec  10.6 MBytes  8.86 Mbits/sec  7598  sender
  0.00-10.50  sec  10.6 MBytes  8.44 Mbits/sec  7598  receiver 
```

For some reason the esp_hosted RX is always half the TX.

robert-hh commented:
> For some reason the esp_hosted RX is always half the TX.

From what I observed, it's a matter of the MCU type. For SAMD and MIMXRT I have ports with both NINAW10 and esp_hosted. With SAMD, RX is the same as TX, with MIMXRT RX is half of TX.  I do not know why.

robert-hh commented:
Good Morning Ibrahim. Since your changes to esp_hosted are now merged, I try to pull up the respective SAMD and MIMXRT branches. There seems to be a problem with calling esp_hosted_hal_deinit(), which is called with esp_hosted_hal_init(), but not on a soft reset, unless bluetooth was enabled. But calling esp_hosted_hal_deinit() is required to call esp_hosted_wifi_deinit(). I used to have that in main.c, but it is not called in the C33 main.c.

iabdalkader commented:
I'm not sure I fully understand the issue, in C33 WiFi is not deinitialized on soft-reset, unless Bluetooth was used, in that case `mp_bluetooth_deinit` calls hci_deinit, which is fine because you can't be using both BT + WiFi at the same time. If WiFi is used it's fine to leave it initialized (in fact I wanted that because it's faster this way).

BTW, did you manage to build the esp_hosted firmware ? If you did, can you please share steps to building the firmware ? I saw some recent changes that may improve the performance, but also wanted to try to increase SPI TX/RX buffers see if it helps. I have IDF 5+ and tried to build it, but I think our sdkconfig.defaults is not actually a defaults file, and the build just keeps reverting to default sdkconfig (with `esp32` target).

robert-hh commented:
I have to disable WiFi, because otherwise there are some dangling net events which cannot be handled since SPI is gone. No problem.

I have indeed built the esp_hosted firmware to change the pins for the interface signals. I went through the IDF-config session using `idp.py menuconfig` and ended up after a few attempts with a working sdkconfig. The default files did not make a lot to it. Attached is my sdkconfig. It's for a generic ESP32. Maybe it helps.

[sdkconfig.zip](https://github.com/micropython/micropython/files/12792245/sdkconfig.zip)

robert-hh commented:
b.t.w: I made a small change to mp_bluetooth_hci_controller_init() to shorten the BLE start-up time. Nothing urgen or ciritcal t, so it can  wait until the support for the respective port is merged, or if there are anyhow changes to be made.
```
diff --git a/drivers/esp-hosted/esp_hosted_bthci.c b/drivers/esp-hosted/esp_hosted_bthci.c
index 003054460..25876f039 100644
--- a/drivers/esp-hosted/esp_hosted_bthci.c
+++ b/drivers/esp-hosted/esp_hosted_bthci.c
@@ -122,9 +122,10 @@ int mp_bluetooth_hci_controller_init(void) {
 
     mp_uint_t start = mp_hal_ticks_ms();
     // Skip bootloader messages.
-    while ((mp_hal_ticks_ms() - start) < 2500) {
+    while ((mp_hal_ticks_ms() - start) < 500) {
         if (mp_bluetooth_hci_uart_any()) {
             mp_bluetooth_hci_uart_readchar();
+            start = mp_hal_ticks_ms();
         }
         MICROPY_EVENT_POLL_HOOK
     }
```

iabdalkader commented:
Actually I think this was required because the bootloader's verbose debugging is enabled (at least in the firmware that ships with C33), which is very annoying.

robert-hh commented:
The esp_hosted variant works very well now. With the combination of a Teensy 4.0 and a Adafruit ESP32 breakout board I get with iperf3 3.4 MBit/s for TCP and 6.2 MBit/s for UDP, with SPI at 8Mbit/s!
And the BLE tests also perform well.

iabdalkader commented:
That's great news, close to the official benchmarks on ESP32.

robert-hh commented:
I have a question about the macro `MICROPY_HW_ESP_HOSTED_SHARED_PINS`. Which signals are assumed to be shared when it's defined as 1?

iabdalkader commented:
The CS pin, you asked for this not be changed on deinit.
